src/macho.rs

```diff
@@ -0,0 +1,265 @@
+                        if segname == b"__TEXT" && sectname == b"__text" {
+                            (SectionKind::Text, SymbolKind::Text)
+                        } else if segname == b"__DATA" && sectname == b"__data" {
+                            (SectionKind::Data, SymbolKind::Data)
+                        } else if segname == b"__DATA" && sectname == b"__bss" {
+                            (SectionKind::UninitializedData, SymbolKind::Data)
+                        } else {
+                            (SectionKind::Other, SymbolKind::Unknown)
+                        };
+                    let section_index = section_kinds.len();
+                    section_kinds.push((section_kind, symbol_kind));
+                    section_ends.push(Symbol {
+                        kind: SymbolKind::Section,
+                        section: section_index + 1,
+                        section_kind: Some(section_kind),
+                        global: false,
+                        name: &[],
+                        address: section.addr + section.size,
+                        size: 0,
+                    });
+                } else {
+                    // Add placeholder so that indexing works.
+                    section_kinds.push((SectionKind::Unknown, SymbolKind::Unknown));
+                }
+            }
+        }
+
+        let mut symbols = Vec::new();
+        for sym in self.macho.symbols() {
+            if let Ok((name, nlist)) = sym {
+                // Skip STAB debugging symbols.
+                // FIXME: use N_STAB constant
+                if nlist.n_type & 0xe0 != 0 {
+                    continue;
+                }
+                let n_type = nlist.n_type & mach::symbols::NLIST_TYPE_MASK;
+                let (section_kind, kind) = if n_type == mach::symbols::N_SECT {
+                    if nlist.n_sect == 0 || nlist.n_sect - 1 >= section_kinds.len() {
+                        (None, SymbolKind::Unknown)
+                    } else {
+                        let (section_kind, kind) = section_kinds[nlist.n_sect - 1];
+                        (Some(section_kind), kind)
+                    }
+                } else {
+                    (None, SymbolKind::Unknown)
+                };
+                symbols.push(Symbol {
+                    kind,
+                    section: nlist.n_sect,
+                    section_kind,
+                    global: nlist.is_global(),
+                    name: name.as_bytes(),
+                    address: nlist.n_value,
+                    size: 0,
+                });
+            }
+        }
+
+        {
+            // Calculate symbol sizes by sorting and finding the next symbol.
+            let mut symbol_refs = Vec::with_capacity(symbols.len() + section_ends.len());
+            symbol_refs.extend(symbols.iter_mut().filter(|s| !s.is_undefined()));
+            symbol_refs.extend(section_ends.iter_mut());
+            symbol_refs.sort_by(|a, b| {
```

Comment:
Here in the mach-o version we're sorting, but in the elf version, we aren't. Are elf symbols always already sorted?

We should document the sortedness guarantees we're providing in the doc comment for the trait method.

Reply:
I've add documentation for the order.

We're only sorting references to the symbols, which we don't return. Part of the reason we don't return the sorted list is that the sorted list also contains `section_ends`, which we don't want to return. The other reason is that the original order may be significant. I'm not sure if this applies to Mach-O, but for ELF the original order is needed to be able to determine which symbols the STT_FILE applies to.

The symbol API probably needs to change in future. Currently it's just something that provides what was needed for addr2line. In particular, we may want to provide an API that returns an iterator (so no allocations), but this iterator won't be able to provide symbol sizes for Mach-O, so we'll still need something for that. Also, the plan is to move the symbol map from addr2line to here, so maybe we can do things differently after that.